include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,323 @@
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    const_iterator() noexcept = default;
+
+    const_iterator(const_iterator const&) noexcept = default;
+
+    const_iterator&
+    operator=(const_iterator const&) noexcept = default;
+
+    reference
+    operator*() const noexcept
+    {
+        char c{'\0'};
+        pct_decode_opts opt;
+        opt.plus_to_space = plus_to_space_;
+        pct_decode_unchecked(
+            &c, &c+1, string_view(pos_, end_), opt);
+        return c;
+    }
+
+    const_iterator&
+    operator++() noexcept
+    {
+        BOOST_ASSERT(pos_ != nullptr);
+        pos_ += *pos_ == '%' ? 3 : 1;
+        return *this;
+    }
+
+    const_iterator&
+    operator--() noexcept
+    {
+        BOOST_ASSERT(pos_ != begin_);
+        if (pos_ - begin_ >= 3 &&
+            *(pos_ - 3) == '%')
+            pos_ -= 3;
+        else
+            --pos_;
+        return *this;
+    }
+
+    const_iterator
+    operator++(int) noexcept
+    {
+        auto tmp = *this;
+        ++*this;
+        return tmp;
+    }
+
+    const_iterator
+    operator--(int) noexcept
+    {
+        auto tmp = *this;
+        --*this;
+        return tmp;
+    }
+
+    const char*
+    base()
+    {
+        return pos_;
+    }
+
+    bool
+    operator==(
+        const_iterator const& other) const noexcept
+    {
+        return pos_ == other.pos_;
+    }
+
+    bool
+    operator!=(
+        const_iterator const& other) const noexcept
+    {
+        return !(*this == other);
+    }
+};
+
+constexpr
+decoded_view::const_iterator
+decoded_view::
+begin() const noexcept
+{
+    return {p_, n_, plus_to_space_};
+}
+
+
+constexpr
+decoded_view::const_iterator
+decoded_view::
+cbegin() const noexcept
+{
+    return begin();
+}
+
+constexpr
+decoded_view::const_iterator
+decoded_view::
+end() const noexcept
+{
+    return {p_, n_, plus_to_space_, 0};
+}
+
+constexpr
+decoded_view::const_iterator
+decoded_view::
+cend() const noexcept
+{
+    return {p_, n_, plus_to_space_, 0};
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reverse_iterator
+decoded_view::
+rbegin() const noexcept
+{
+    return const_reverse_iterator(end());
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reverse_iterator
+decoded_view::
+crbegin() const noexcept
+{
+    return rbegin();
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reverse_iterator
+decoded_view::
+rend() const noexcept
+{
+    return const_reverse_iterator(begin());
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reverse_iterator
+decoded_view::
+crend() const noexcept
+{
+    return rend();
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reference
+decoded_view::
+front() const
+{
+    BOOST_ASSERT( !empty() );
+    return *begin();
+}
+
+BOOST_URL_CXX17_CONSTEXPR inline
+decoded_view::const_reference
+decoded_view::
+back() const
+{
+    BOOST_ASSERT( !empty() );
+    return *rbegin();
```

Comment:
why would you do this? the `reverse_iterator` "support" is already of questionable value and try-hard, and now we are making the compiler go through the work of instantiating `*rbegin()`? Why not just return `*--end()`?